tools/experiments/idle_ext_pmt.py

```diff
@@ -0,0 +1,255 @@
+        ARTIQ python can recognize its type.
+
+        Returns:
+            TTL changes. The first element is always a placeholder.
+        """
+        ttl_changes = self.cxn.artiq.get_ttl_change_queues()
+        to_kernel = [(-1, 0.)]
+        for kk in ttl_changes:
+            index = self.devices.ttl_outs.index(kk[0])
+            to_kernel.append((index, kk[1]))
+        return to_kernel
+
+    @kernel(flags={"fast-math"})
+    def update_dds(self, index: TInt32, index_repump: TInt32, attribute: TStr, value: TFloat):
+        """Sets DDS value."""
+        device = self.ad9910s[index]
+        if attribute == "frequency":
+            freq, phase, amp = device.get()
+            self.core.break_realtime()
+            delay_mu(self.dds_set_delay_mu)
+            device.set(value, phase, amp)
+            self.core.break_realtime()
+        elif attribute == "phase":
+            freq, phase, amp = device.get()
+            self.core.break_realtime()
+            delay_mu(self.dds_set_delay_mu)
+            device.set(freq, value, amp)
+            self.core.break_realtime()
+        elif attribute == "amplitude":
+            freq, phase, amp = device.get()
+            self.core.break_realtime()
+            delay_mu(self.dds_set_delay_mu)
+            device.set(freq, phase, value)
+            self.core.break_realtime()
+        elif attribute == "attenuation":
+            # get_att_mu() required to correctly set the atts of other DDSes of the same urukul.
+            device.get_att_mu()
+            self.core.break_realtime()
+            device.set_att(value)
+            self.core.break_realtime()
+        elif attribute == "state":
+            is_on = value > 0.
+            if is_on:
+                device.sw.on()
+            else:
+                device.sw.off()
+            if index_repump >= 0:
+                self.repump_aom_states[index_repump] = value
+            self.core.break_realtime()
+
+    @kernel(flags={"fast-math"})
+    def update_ttl(self, index: TInt32, value: TFloat):
+        """Sets TTL value."""
+        device = self.ttl_outs[index]
+        if value > 0.:
+            device.on()
+        else:
+            device.off()
+        self.core.break_realtime()
+
+    @rpc
+    def switch_repump_aom_states(self, turn_off):
+        for kk in range(len(self.repump_aoms)):
+            # if the repump AOM is off, don't turn on and off the AOM.
+            # the repump AOM stays off for both differential high and low counting periods.
+
+            if self.repump_aom_states[kk] > 0.:
+                if turn_off:
+                    self.repump_aoms[kk].sw.off()
+                else:
+                    self.repump_aoms[kk].sw.on()
+
+    @rpc
+    def get_max_cycle(self):
+        """Gets PMT counting interval.
+
+        This is the max wait time for differential mode. If PMT is not on,
+        this defaults to 0."""
+        try:
+            interval = self.cxn.pmt_arduino.get_interval()
```

Comment:
Actually since the code already checks the interval. It should probably check whether it is in differential mode too. If no, it does not need to do anything. If yes, it should wait for first the first rising pulse for maximumly 2x interval (to make sure that the first rising pulse is caught).

Reply:
should it only wait 2x for the first rising pulse, or all of them?

Reply:
I think we can wait for 2x for all of them (to account for small timing differences).